src/qml/qmlwaveformrenderer.cpp

```diff
@@ -0,0 +1,243 @@
+                pMark->visibilityControl(),
+                pMark->durationTextLocation()));
+    }
+    return QmlWaveformRendererFactory::Renderer{renderer, renderer};
+}
+
+#ifdef __STEM__
+QmlWaveformRendererFactory::Renderer QmlWaveformRendererStem::create(
+        WaveformWidgetRenderer* waveformWidget) const {
+    auto* renderer = new WaveformRendererStem(
+            waveformWidget, m_position);
+
+    renderer->setAllChannelVisualGain(m_gainAll);
+    connect(this,
+            &QmlWaveformRendererStem::gainAllChanged,
+            renderer,
+            &WaveformRendererStem::setAllChannelVisualGain);
+    return QmlWaveformRendererFactory::Renderer{renderer, renderer};
+}
+#endif
+
+QmlWaveformRendererFactory::Renderer QmlWaveformRendererMark::create(
+        WaveformWidgetRenderer* waveformWidget) const {
+    VERIFY_OR_DEBUG_ASSERT(!!m_untilMark) {
+        return QmlWaveformRendererFactory::Renderer{};
+    }
+    auto* renderer = new WaveformRenderMark(waveformWidget,
+            ::WaveformRendererAbstract::Play);
+
+    renderer->setFgPlayColor(m_playMarkerColor);
+    renderer->setBgPlayColor(m_playMarkerBackground);
+
+    renderer->setUntilMarkShowBeats(m_untilMark->showTime());
+    renderer->setUntilMarkShowTime(m_untilMark->showBeats());
+    renderer->setUntilMarkAlign(static_cast<Qt::Alignment>(m_untilMark->align()));
+    renderer->setUntilMarkTextSize(m_untilMark->textSize());
+    renderer->setUntilMarkTextHeightLimit(m_untilMark->textHeightLimit());
+
+    connect(m_untilMark.get(),
+            &QmlWaveformUntilMark::showTimeChanged,
+            renderer,
+            &WaveformRenderMark::setUntilMarkShowTime);
+    connect(m_untilMark.get(),
+            &QmlWaveformUntilMark::showBeatsChanged,
+            renderer,
+            &WaveformRenderMark::setUntilMarkShowBeats);
+    connect(m_untilMark.get(),
+            &QmlWaveformUntilMark::alignChanged,
+            renderer,
+            &WaveformRenderMark::setUntilMarkAlign);
+    connect(m_untilMark.get(),
+            &QmlWaveformUntilMark::textSizeChanged,
+            renderer,
+            &WaveformRenderMark::setUntilMarkTextSize);
+
+    connect(this,
+            &QmlWaveformRendererMark::playMarkerColorChanged,
+            renderer,
+            &WaveformRenderMark::setFgPlayColor);
+    connect(this,
+            &QmlWaveformRendererMark::playMarkerBackgroundChanged,
+            renderer,
+            &WaveformRenderMark::setBgPlayColor);
+
+    int priority = 0;
+    for (auto* pMark : m_marks) {
+        renderer->addMark(WaveformMarkPointer(new WaveformMark(
+                waveformWidget->getGroup(),
+                pMark->control(),
+                pMark->visibilityControl(),
+                pMark->textColor(),
+                pMark->align(),
+                pMark->text(),
+                pMark->pixmap(),
+                pMark->icon(),
+                pMark->color(),
+                --priority)));
```

Comment:
Is this really supposed to go 0, -1, -2, ...?

Also, I would be in favor of splitting the increment/decrement step onto a separate line for readability.

Reply:
This was copied from the renderer setup logic used by the legacy skin parser so I cannot do much on what is expected there. Sure, happy to make that on two lines

Reply:
Added a comment above to explain where this logic comes from